atcoder/maxflow.py

```diff
@@ -0,0 +1,209 @@
+            re.cap
+        )
+
+    def edges(self) -> List[Edge]:
+        return [self.get_edge(i) for i in range(len(self._edges))]
+
+    def change_edge(self, i: int, new_cap: int, new_flow: int):
+        assert 0 <= i < len(self._edges)
+        assert 0 <= new_flow <= new_cap
+        e = self._edges[i]
+        e.cap = new_cap - new_flow
+        e.rev.cap = new_flow
+
+    def flow(self, s: int, t: int, flow_limit: Optional[int] = None) -> int:
+        assert 0 <= s < self._n
+        assert 0 <= t < self._n
+        assert s != t
+        if flow_limit is None:
+            flow_limit = sum(e.cap for e in self._g[s])
+
+        current_edge = [0] * self._n
+        level = [0] * self._n
+
+        def fill(arr: List[int], value: int):
+            for i in range(len(arr)):
+                arr[i] = value
+
+        def bfs() -> bool:
+            fill(level, self._n)
+            queue = []
+            q_front = 0
+            queue.append(s)
+            level[s] = 0
+            while q_front < len(queue):
+                v = queue[q_front]
+                q_front += 1
+                next_level = level[v] + 1
+                for e in self._g[v]:
+                    if e.cap == 0 or level[e.dst] <= next_level:
+                        continue
+                    level[e.dst] = next_level
+                    if e.dst == t:
+                        return True
+                    queue.append(e.dst)
+            return False
+
+        def dfs(lim) -> int:
+            stack = []
+            edge_stack = []
+            stack.append(t)
+            while stack:
+                v = stack[-1]
+                if v == s:
+                    flow = min(lim, min(e.cap for e in edge_stack))
+                    for e in edge_stack:
+                        e.cap -= flow
+                        e.rev.cap += flow
+                    return flow
+                next_level = level[v] - 1
+                while current_edge[v] < len(self._g[v]):
+                    e = self._g[v][current_edge[v]]
+                    re = e.rev
+                    if level[e.dst] != next_level or re.cap == 0:
+                        current_edge[v] += 1
+                        continue
+                    stack.append(e.dst)
+                    edge_stack.append(re)
+                    break
+                else:
+                    stack.pop()
+                    if edge_stack:
+                        edge_stack.pop()
+                    level[v] = self._n
+            return 0
+
+        flow = 0
+        while flow < flow_limit:
+            if not bfs():
+                break
+            fill(current_edge, 0)
+            while flow < flow_limit:
+                f = dfs(flow_limit - flow)
+                flow += f
+                if f == 0:
+                    break
+        return flow
+
+    def min_cut(self, s: int) -> List[bool]:
+        visited = [False] * self._n
+        stack = [s]
+        visited[s] = True
```

Comment:
Note: The original impl had `visited[v] = True` after `stack.pop()` instead of this.